src/python/pants/backend/java/bsp/rules.py

```diff
@@ -0,0 +1,221 @@
+    )
+
+
+@rule
+async def bsp_resolve_all_java_build_targets(
+    _: JavaBSPBuildTargetsRequest,
+    all_java_targets: AllJavaTargets,
+    bsp_context: BSPContext,
+) -> BSPBuildTargets:
+    if LANGUAGE_ID not in bsp_context.client_params.capabilities.language_ids:
+        return BSPBuildTargets()
+    build_targets = await MultiGet(
+        Get(BuildTarget, ResolveJavaBSPBuildTargetRequest(tgt)) for tgt in all_java_targets
+    )
+    return BSPBuildTargets(targets=tuple(build_targets))
+
+
+# -----------------------------------------------------------------------------------------------
+# Javac Options Request
+# See https://build-server-protocol.github.io/docs/extensions/java.html#javac-options-request
+# -----------------------------------------------------------------------------------------------
+
+
+class JavacOptionsHandlerMapping(BSPHandlerMapping):
+    method_name = "buildTarget/javacOptions"
+    request_type = JavacOptionsParams
+    response_type = JavacOptionsResult
+
+
+@dataclass(frozen=True)
+class HandleJavacOptionsRequest:
+    bsp_target_id: BuildTargetIdentifier
+
+
+@dataclass(frozen=True)
+class HandleScalacOptionsResult:
+    item: JavacOptionsItem
+
+
+@rule
+async def handle_bsp_scalac_options_request(
+    request: HandleJavacOptionsRequest,
+    build_root: BuildRoot,
+) -> HandleScalacOptionsResult:
+    wrapped_target = await Get(WrappedTarget, AddressInput, request.bsp_target_id.address_input)
+    coarsened_targets = await Get(CoarsenedTargets, Addresses([wrapped_target.target.address]))
+    assert len(coarsened_targets) == 1
+    coarsened_target = coarsened_targets[0]
+    resolve = await Get(CoursierResolveKey, CoarsenedTargets([coarsened_target]))
+    output_file = compute_output_jar_filename(coarsened_target)
+
+    return HandleScalacOptionsResult(
+        JavacOptionsItem(
+            target=request.bsp_target_id,
+            options=(),
+            classpath=(
+                build_root.pathlib_path.joinpath(
+                    f".pants.d/bsp/jvm/resolves/{resolve.name}/lib/{output_file}"
+                ).as_uri(),
+            ),
+            class_directory=build_root.pathlib_path.joinpath(
+                f".pants.d/bsp/jvm/resolves/{resolve.name}/classes"
+            ).as_uri(),
+        )
+    )
+
+
+@rule
+async def bsp_javac_options_request(request: JavacOptionsParams) -> JavacOptionsResult:
+    results = await MultiGet(
+        Get(HandleScalacOptionsResult, HandleJavacOptionsRequest(btgt)) for btgt in request.targets
+    )
+    return JavacOptionsResult(items=tuple(result.item for result in results))
+
+
+# -----------------------------------------------------------------------------------------------
+# Compile Request
+# -----------------------------------------------------------------------------------------------
+
+
+@dataclass(frozen=True)
+class JavaBSPCompileFieldSet(BSPCompileFieldSet):
+    required_fields = (JavaSourceField,)
+    source: JavaSourceField
```

Comment:
Rather than creating a new FieldSet, you can use https://github.com/pantsbuild/pants/blob/16306ed3d182b93d2d53d9685389201066275840/src/python/pants/backend/java/target_types.py#L34-L38 directly probably.

Reply:
That might be unideal coupling. The theory of FieldSets was to have one FieldSet per use case.

Reply:
It's a bit premature to optimize this away. Also, I will likely want to expose the target generators and not the file-level targets as BSP build targets. Keeping the field set separate enables exploration into that use case.